### ISSUE TITLE
 InterFuser (26.3M) trained on the  tiny dataset(provided) with/without plankd. But I obtained abnormal driving score.

### ISSUE THREAD

**Issue by c122-ode** [tulerfeng/PlanKD#3]
I traind InterFuser (26.3M) on the  tiny dataset you provided. But the results are abnormal. You said that you get the 36.52 / 25.54 driving score with / without PlanKD on Town05 Short. So could you tell me what parameters used in the model training.

I trained model with below settings: 

**teacher ckpt** is what you provided: https://drive.google.com/file/d/19IOwQoNtCMg36aWsshicJHgH-FtcWdDz/view?usp=sharing
**train script**: interfuser/scripts/train_plankd.sh  and  interfuser/scripts/train.sh 
change the parameters below:
`--train-towns 1 2 3 4 6 7 10  --val-towns 5 --train-weathers 0 2 3 5  --val-weathers 0 2 3 5`

**Results**
**with plankd:**
`"values": [
        "35.117",
        "66.970",
        "0.513",
        "0.382",
        "0.000",
        "1.562",
        "4.083",
        "0.000",
        "0.424",
        "0.000",
        "2.493",
        "0.000"
    ]`

**without plankd:**
` "values": [
        "34.606",
        "84.381",
        "0.397",
        "0.722",
        "1.359",
        "1.208",
        "2.119",
        "0.000",
        "1.298",
        "0.000",
        "1.016",
        "0.000"
    ]`

**Comment by c122-ode** [comment]
我发现在 interfuser/train.py 
 train_one_epoch() 应该加入model.teacher_model.eval()
model.IB 参数的更新在 Plankd的train_IB进行，为啥train.py 中 还加入到 other_weights中

**Comment by tulerfeng** [comment]
您好，感谢您对我们工作的兴趣。

我重新训练和测试了一下，用了plankd还是能够稳定提升的。 我统一采用第10个epoch（如论文附录所言）得到的权重进行测试。

即，对于with plankd的方法，evaluation文件设置的测试模型是checkpoint_student-9.pth.tar，对于without plankd的方法，evaluation文件设置的测试模型是checkpoint-9.pth.tar

我三次测试的结果如下：

with plankd: 32.595, 38.358, 33.017

without plankd: 30.881, 29.604, 28.483

这个carla的评估测试可能随机性比较大，您可以多测试几次试试看。

我得到的模型权重文件，以及所有的测试结果文件，上传到了这个链接里面，您可以参考一下：https://drive.google.com/file/d/1bnmMNs604dlW6wmBPbgLtfqv-Hrd1DUC/view?usp=sharing

**Comment by tulerfeng** [comment]
对于‘train_one_epoch() 应该加入model.teacher_model.eval()’这个问题，也许加入这个会更好，不过应该对我们的方法的有效性影响不大

对于‘为啥train.py 中 还加入到 other_weights中’这个问题，因为我们想用类似端到端的方法去训练，希望决策的损失也可以优化到前面的一些参数，所以加入了这个

**Comment by c122-ode** [comment]
非常感谢您的回复！还有几个问题，您有时间请看一下。非常感谢！
1. 我是根据eval_l1_error 最小的epoch 选择的要测试的epoch。
with plankd: epoch12   (不知道为什么，参数设置的epochs 是10，但是训练完成后实际上是13个epoch)
without plankd epoch8
如果按照epoch-9测试，without plankd 和您的结果接近。
2. 参数中score_penalty 和 score_route您会关注吗，我发现虽然score_composed 高，但是有时候score_route却比较低。
3. 希望决策的损失也可以优化到前面的一些参数：那么model.IB 的参数实际上在一次处理中 会进行2次参数更新。这种是第一次见哦，这种很神奇！
4. wp_attention的输入都是原始数据中的，实际上不是模型去产生的一些信息，那这个产生的是2个损失 entropy, rank_loss是用来更新什么参数的呢
`if self.base_model == 'interfuser':
                att, entropy, rank_loss = self.wp_attention(x['gt'], x['bev'], x['measurements'], x['actor_pos'], x['actor_cnt'])`
wp_attention 应该是论文中的 Safety-aware Waypoint-attentive Distillation，rank_loss 是Safety-aware ranking loss，entropy是Waypoint-attentive distillation, 但是 从代码看 这个损失的计算中没有教师的任何信息哇

**Comment by tulerfeng** [comment]
您好：
1. 因为interfuser的代码设置里面，有一些warm up的epoch，所以会比10更多。对于大数据量的训练，我们是训练到第10个epoch就直接手动中止训练了（在原始数据量下，一个epoch需要训练一天），不会训练后面的3个epoch
2. 因为driving score是主要的指标，所以我们主要关注这个driving score，对于score_penalty 和 score_route，我们是希望penalty尽量高一些就行，因为自动驾驶这边安全性是最重要的，就算route完成率比较低，也不希望撞车之类的
3. 这个地方可能确实代码没写太好
4.  entropy, rank_loss是用来优化论文里面的BEV Encoder和Waypoint Encoder的参数的（同时，最后的决策损失，也会优化到这两个encoder），不需要用到教师模型的信息（教师模型的信息，只在前面那个IB模块用到）。但是，如果后面有比收集数据的 expert model更好的教师模型，这里的输入路点，也可以用教师模型的。

**Comment by c122-ode** [comment]
十分感谢您的回复，受益匪浅！
